File: opt/coloring/solver.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
from copy import copy
import profile


def checkConfilct(node, color, nodes_color, d):
    l = len(nodes_color)
    for w in d[node]:
        if w < l and color == nodes_color[w]:
            return False
    return True

def colorSpace(node, nodes_color, d):
    s = set(nodes_color[:node])
    for i in range(node):
        if d[node][i] == 1 and nodes_color[i] in s:
            s.remove(nodes_color[i])

    # flag = True
    # for i in range(node, len(d)):
    #     if d[node][i] == 1:
    #         flag = False
    #         break
    #
    # if flag and len(s) > 0:
    #     s = [s.pop()]
    return s

def solveIt(inputData):
    # Modify this code to run your optimization algorithm

    # parse the input
    lines = inputData.split('\n')

    firstLine = lines[0].split()
    nodeCount = int(firstLine[0])
    edgeCount = int(firstLine[1])

    edges = []
    for i in range(1, edgeCount + 1):
        line = lines[i]
        parts = line.split()
        edges.append((int(parts[0]), int(parts[1])))

    # build a trivial solution
    # every node has its own color
    solution = range(0, nodeCount)

    d = [[0 for i in range(nodeCount)] for j in range(nodeCount)]

    for v,w in edges:
        d[v][w] = 1
        d[w][v] = 1

    s = []
    s.append((1, 0, 0))
    min_colors = nodeCount + 1
    min_nodes_color = []
    nodes_color = [0 for i in range(nodeCount)]
    while len(s) != 0:
        colors_count, node, node_color = s.pop()
        if colors_count < min_colors:
            nodes_color[node] = node_color
            if node == nodeCount - 1:
                min_colors = colors_count
                min_nodes_color = copy(nodes_color)
            else:
                s.append((colors_count+1, node+1, colors_count))
                for color in colorSpace(node+1, nodes_color, d):
                    s.append((colors_count, node+1, color))
                # for color in range(colors_count-1, -1, -1):
                #     if checkConfilct(node, color, nodes_color, d):
                #         temp = copy(nodes_color)
                #         temp.append(color)
                #         s.append((colors_count, temp))

    solution = min_nodes_color

    nodeCount = min_colors
    # prepare the solution in the specified output format
    outputData = str(nodeCount) + ' ' + str(1) + '\n'
    outputData += ' '.join(map(str, solution))

    return outputData


def ps():
    if len(sys.argv) > 1:
        fileLocation = sys.argv[1].strip()
        inputDataFile = open(fileLocation, 'r')
        inputData = ''.join(inputDataFile.readlines())
        inputDataFile.close()
        print solveIt(inputData)
    else:
        print 'This test requires an input file.  Please select one from the data directory. (i.e. python solver.py ./data/gc_4_1)'



import sys

if __name__ == '__main__':
    profile.run('ps()')
    #ps()

File: opt/coloring/Solver.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * Created with IntelliJ IDEA.
 * User: Administrator
 * Date: 13-6-25
 * Time: ÏÂÎç9:28
 * To change this template use File | Settings | File Templates.
 */
public class Solver {

    public static void main(String[] args) {
        try {
            solve(args);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public final void colorSpace(int node, int nodeColor, int[][] edges){

    }

    public static void solve(String[] args) throws IOException {
        String fileName = null;

        // get the temp file name
        for(String arg : args){
            if(arg.startsWith("-file=")){
                fileName = arg.substring(6);
            }
        }
        if(fileName == null)
            return;

        // read the lines out of the file
        List<String> lines = new ArrayList<String>();

        BufferedReader input =  new BufferedReader(new FileReader(fileName));
        try {
            String line = null;
            while (( line = input.readLine()) != null){
                lines.add(line);
            }
        }
        finally {
            input.close();
        }


        // parse the data in the file
        String[] firstLine = lines.get(0).split("\\s+");
        int nodeCount = Integer.parseInt(firstLine[0]);
        int edgeCount = Integer.parseInt(firstLine[1]);

        int[][] edges = new int[nodeCount][nodeCount];
        int v, w;
        for (int i = 1; i < edgeCount + 1; i++){
            String line = lines.get(i);
            String[] parts = line.split("\\s+");
            v = Integer.parseInt(parts[0]);
            w = Integer.parseInt(parts[1]);
            edges[v][w] = 1;
            edges[w][v] = 1;


        }

        Stack<State> s = new Stack<State>();
        s.push(new State(1, 0, 0));

        int minColorsNum = nodeCount + 1;
        int[] minNodesColor = new int[nodeCount];
        int[] nodesColor = new int[nodeCount];
        int node, color, colorCount, newNode;
        int end = nodeCount - 1;
        boolean flag;
        while (!s.empty()){
            State state = s.pop();
            node = state.Node;
            if (state.ColorCount < minColorsNum){
                color = state.NodeColor;
                nodesColor[node] = color;
                if (node == end){
                    minColorsNum = state.ColorCount;
                    minNodesColor = nodesColor.clone();
                }
                else{
                    colorCount = state.ColorCount;
                    newNode = node + 1;
                    s.push(new State(colorCount + 1, newNode, colorCount));
                    for (int i= 0; i < colorCount; i++){
                         flag = true;
                         for (int j = 0; j <= node; j++){
                              if (edges[newNode][j] == 1 && nodesColor[j] == i){
                                  flag = false;
                                  break;
                              }
                         }

                        if (flag){
                            s.push(new State(colorCount, newNode, i));
                        }
                    }
                }
            }
        }
	minColorsNum -= 1;
        System.out.println(minColorsNum+" 1");
        for(int i=0; i < nodeCount; i++){
            System.out.print(minNodesColor[i]+" ");
        }
        System.out.println("");
    }
}
